File: src/app/models/map/game-map.ts
import { Character } from '@models/characters/character';
import {CellType} from './cell-type';
import {Dot} from '../interfaces';

export class GameMap {
    private grid: CellType[][];
    
    constructor(mapData: number[][]) {
        this.grid = mapData.map(row => row.map(cell => cell as CellType));
    }
    
    getCell(x: number, y: number): CellType | null {
        return this.isValidPosition(x, y) ? this.grid[y][x] : null;
    }
    
    setCell(x: number, y: number, type: CellType): void {
        if (this.isValidPosition(x, y)) this.grid[y][x] = type;
    }
    
    isWalkable(character: Character, x: number, y: number): boolean {
        const cell = this.getCell(x, y) ?? CellType.Wall; 

        if (cell === CellType.Door) {
            return character.canPassDoor();
        }

        return cell !== CellType.Wall;
    }
    
    private isValidPosition(x: number, y: number): boolean {
        return x >= 0 && y >= 0 && y < this.height && x < this.width;
    }
    
    get width(): number {
        return this.grid[0]?.length ?? 0;
    }
    
    get height(): number {
        return this.grid.length;
    }
    
    get dots(): Dot[] {
        return this.grid
            .flatMap((row, y) => row.map((cell, x) => ({cell, x, y})))
            .filter(({cell}) => [CellType.Dot, CellType.PowerPellet].includes(cell))
            .map(({x, y, cell}) => ({gridX: x, gridY: y, type: cell}));
    }
    
    get walls(): {gridX: number, gridY: number}[] {
        return this.grid
            .flatMap((row, y) => row.map((cell, x) => ({cell, x, y})))
            .filter(({cell}) => cell === CellType.Wall)
            .map(({x, y}) => ({gridX: x, gridY: y}));
    }

    get door(): {gridX: number, gridY: number} | null {
        for (let y = 0; y < this.grid.length; y++) {
            for (let x = 0; x < this.grid[y].length; x++) {
                if (this.grid[y][x] === CellType.Door) {
                    return {gridX: x, gridY: y};
                }
            }
        }
        return null;
    }
    
    drawWalls(ctx: CanvasRenderingContext2D, cellSize: number): void {
        ctx.fillStyle = '#0000FF';
        ctx.fillRect(0, 0, this.width * cellSize, this.height * cellSize);
        
        ctx.fillStyle = '#000000';
        this.grid.forEach((row, y) => {
            row.forEach((cell, x) => {
                if (cell !== CellType.Wall) {
                    ctx.fillRect(
                        x * cellSize,
                        y * cellSize,
                        cellSize,
                        cellSize
                    );
                }
            });
        });
    }
    
    drawDots(ctx: CanvasRenderingContext2D, cellSize: number): void {
        ctx.fillStyle = '#FFFFFF';
        this.dots.forEach(({ gridX, gridY, type }) => {
            const x = gridX * cellSize + cellSize / 2;
            const y = gridY * cellSize + cellSize / 2;
            const radius = type === CellType.Dot ? 2 : 6;
            
            ctx.beginPath();
            ctx.arc(x, y, radius, 0, Math.PI * 2);
            ctx.fill();
        });
    }

    drawDoor(ctx: CanvasRenderingContext2D, cellSize: number): void {
        const door = this.door;
        if (door) {
            ctx.fillStyle = '#FF69B4';
            const doorWidth = cellSize;
            const doorHeight = cellSize / 5;
            const offsetX = (cellSize - doorWidth) / 2;
            ctx.fillRect(
                door.gridX * cellSize + offsetX,
                door.gridY * cellSize,
                doorWidth,
                doorHeight
            );
        }
    }
}

File: server/models/map/game_map.py
from models.interfaces import CellType, Dot
from typing import List, Optional, Tuple

class GameMap:
    def __init__(self, map_data: List[List[int]]):
        self.grid = [[CellType(cell) for cell in row] for row in map_data]
        self.width = len(self.grid[0]) if self.grid else 0
        self.height = len(self.grid)
    
    def get_cell(self, x: int, y: int) -> Optional[CellType]:
        if 0 <= x < self.width and 0 <= y < self.height:
            return self.grid[y][x]
        return None
    
    def set_cell(self, x: int, y: int, cell_type: CellType):
        if 0 <= x < self.width and 0 <= y < self.height:
            self.grid[y][x] = cell_type
    
    def is_walkable(self, x: int, y: int, can_pass_door: bool = False) -> bool:
        cell = self.get_cell(x, y)
        
        if cell == CellType.Door:
            return can_pass_door
            
        return cell != CellType.Wall
    
    def get_dots(self) -> List[Dot]:
        dots = []
        
        for y in range(self.height):
            for x in range(self.width):
                cell = self.grid[y][x]
                
                if cell in [CellType.Dot, CellType.PowerPellet]:
                    dots.append(Dot(x, y, cell))
                    
        return dots
    
    def get_teleport_points(self) -> List[Tuple[int, int]]:
        points = []
        
        for y in range(self.height):
            for x in [0, self.width - 1]:
                cell = self.get_cell(x, y)
                
                if cell in [CellType.Empty, CellType.Dot, CellType.PowerPellet]:
                    points.append((x, y))
                    
        return points